lib/AST/ASTScope.cpp

```diff
@@ -1777,7 +1777,10 @@ DeclContext *ASTScope::getInnermostEnclosingDeclContext() const {
   for (const ASTScope *scope = this; ; scope = scope->getParent()) {
     if (auto dc = scope->getDeclContext()) return dc;
   }
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunreachable-code"
   llvm_unreachable("Top-most scope is a declaration context");
+#pragma clang diagnostic pop
```

Comment:
Oh, 3.8 has __builtin_unreachable. There are a lot of double-negatives involved in what's going on here; the issue has nothing to do with that being a call to llvm_unreachable, it could just as well be a call to printf. This is in an NDEBUG build, which means llvm_unreachable is actually a live (diagnostics-reporting, noreturn) function call.

The issue is that it's happening in a dead-code context, and -Wunreachable-code is complaining (correctly) that it's dead code! Which it is. That llvm_unreachable there really _is_ unreachable, not just in a "check my logic" sense, but in a language-definition sense. It's a loop that only has a return-edge, not an successor-block edge. Control can't get to the llvm_unreachable call, period.

Unfortunately on the clang we're shipping in Xcode, it seems -Wunreachable-code currently does not imply -Wunreachable-code-return (or other sub-flavours of -Wunreachable-code-aggressive); I am not sure why, I'll look a bit more into that now. On non-apple compilers such as the one on ubuntu, -Wunreachable-code detects and warns about this dead-code context.

In 2ba762a I removed such a definitely-trivially-dead call to llvm_unreachable; this is really an additional example of that, but I figured in this case I'd leave it here and pragma-out the warning in case some compiler isn't clever enough to realize it's statically dead. I'm not sure which forms of dead-code-detection one can assume across all compilers these days. If you prefer I can also just delete it here, and we can re-add it only if some other compiler breaks (I just tested gcc and some versions at least seem fine with it; not sure about MSVC).